pkg/generate/app/cmd/newapp.go

```diff
@@ -219,30 +221,80 @@ func (c *AppConfig) SetOpenShiftClient(osclient client.Interface, OriginNamespac
 	}
 }
 
+func (c *AppConfig) addEnvironmentArguments(s string) bool {
+	rc := cmdutil.IsEnvironmentArgument(s)
+	if rc {
+		glog.V(2).Infof("treating %s as possible environment argument\n", s)
+		c.Environment = append(c.Environment, s)
+	}
+	return rc
+}
+
+func (c *AppConfig) addSourceArguments(s string) bool {
+	remote, rerr := app.IsRemoteRepository(s)
+	local, derr := app.IsDirectory(s)
+
+	if remote || local {
+		glog.V(2).Infof("treating %s as possible source repo\n", s)
+		c.SourceRepositories = append(c.SourceRepositories, s)
+		return true
+	}
+
+	if rerr != nil {
+		c.ArgumentClassificationErrors["Remote access"] = rerr
+	}
+
+	if derr != nil {
+		c.ArgumentClassificationErrors["Local access"] = derr
+	}
+
+	return false
+}
+
+func (c *AppConfig) addComponentArguments(s string) bool {
+	err := app.IsComponentReference(s)
+	if err == nil {
+		glog.V(2).Infof("treating %s as a component ref\n", s)
+		c.Components = append(c.Components, s)
+		return true
+	}
+	c.ArgumentClassificationErrors["Component reference processing"] = err
```

Comment:
i'd like to see another term used here that will have meaning to the user.  I think this is essentially checking if the value is either:

1) an image
2) an imagestreamtag
3) a template that exists in the cluster(?)

if it weren't for 3 this would be easier to fix.